### DISCUSSION TITLE
Suggested value for `video_burst_bits`

### DISCUSSION THREAD

**Discussion by maximilianfridrich** [baresip/baresip#3219] [General]
Since #2482 BareSIP is capable of paced video sending (config setting `video_burst_bits`) in order to not cause issues like network congestion (among other issues). If `video_burst_bits` is not set at all, video sending is very jittery and messy.

What we were wondering is what would be an appropriate setting for `video_burst_bits` in relation to `video_bitrate` (or `video_sendrate`)? Does anyone have any insight or suggestions into this? Maybe @sreimers since you implemented this feature, what would be your suggestion?

E.g. let's say `video_bitrate=512000`, what would be an appropriate (default?) value for `video_burst_bits`?

Reference regarding paced video sending:
- https://chromium.googlesource.com/external/webrtc/+/master/modules/pacing/g3doc/index.md
- https://ffmpeg.org/ffmpeg-protocols.html#udp

PS: We did some analysis of the code and the video pacing algorithm. The `video_burst_bits` value effectively determines how "late" a frame from the encoder can be before the bitrate calculation is reset. For example, if `video_bitrate` and `video_burst_bits` are equal, that effectively means that a video frame from the encoder can be up to 1 second late and the video sending algorithm will just send many frames and temporarily overshoot the bitrate (because for the last second no data was sent - but this can potentially lead to congestion). Smaller values mean this calculation is reset even if the frame is just a little bit late.

Here is an example graph we made to analyze this (`video_src` is `avformat,file.mp4`). bitrate=512 kbit/s, burst_bits=50000 bit, burst_interval=50000bit/512000bit/s=0.0976s => once a frame is more than ~100ms later than the next target time, the pacing is reset and we throttle to prevent bursts.

![video_100ms](https://github.com/user-attachments/assets/c130888f-cc63-447d-a515-a998d09f391d)

**Comment by sreimers** [comment]
Here are some production values:
```c
"video_bitrate		2500000\n"
"video_sendrate		10000000\n" 
"video_burst_bit	1000000\n"
```
To allow a bigger/smoother overshot, I set a different sendrate. And it should help to use a video codec with a good CBR setting, otherwise video codecs can easily overshot.

**Comment by sreimers** [comment]
![image](https://github.com/user-attachments/assets/7c84dde3-f7ca-4499-8001-c1a87ddd6630)

These values look strange. are these frame or packet delays? `max_delay` for a packet should be much smaller:

```c
max_delay = 1280 * 8 * 1000000LL / 512000 + 1; /* 512000 bit/s */
max_delay = 20001; /* ~20ms */
```

`video_sendrate` has to be always higher than codec bitrate, since rtp packetization adds a small overhead.

**Comment by maximilianfridrich** [comment]
Thank you for the quick response and suggestions!

> These values look strange. are these frame or packet delays?

These are the frame delays, the naming of the plots is a little confusing. Every dot in that graph shows when a frame from the encoder was later than the allowed `burst_interval` of ~100ms. We see that the frame sizes were much smaller when the frames were late.
